File: observability/examples/chatbot-python/chatbot.py
"""Multimodal chatbot demo using the Theta Observability Python SDK + real Anthropic Claude.

Walks through 5 conversation turns where the user uploads an image and chats
about it. Each turn = one Trace with two Steps (retrieval + llm), and the
image is attached to the trace.

Run:
    THETA_API_KEY=tk_... THETA_PROJECT=proj_... \
    THETA_BASE_URL=http://localhost:8080 \
    ANTHROPIC_API_KEY=sk-ant-... \
    python observability/examples/chatbot-python/chatbot.py
"""

from __future__ import annotations

import base64
import os
import random
import time
from pathlib import Path

from anthropic import Anthropic
from theta_observability import TraceClient

HERE = Path(__file__).parent
IMAGES = sorted((HERE / "images").glob("*.png"))
MODEL = "claude-opus-4-5"  # latest Opus alias

USER_MESSAGES = [
    "Here's an image — what do you see?",
    "Describe the dominant colors more precisely.",
    "Could this be a sunset, ocean, or forest?",
    "Summarise everything we've discussed so far in two lines.",
    "Write a one-line caption I could use on Instagram.",
]


def call_claude(client: Anthropic, image_bytes: bytes, history: list[dict]) -> tuple[str, dict]:
    img_b64 = base64.b64encode(image_bytes).decode()
    messages = []
    for h in history:
        messages.append({"role": h["role"], "content": [{"type": "text", "text": h["text"]}]})
    # last user message also carries the image
    if messages and messages[-1]["role"] == "user":
        messages[-1]["content"].insert(
            0,
            {"type": "image", "source": {"type": "base64", "media_type": "image/png", "data": img_b64}},
        )
    resp = client.messages.create(
        model=MODEL,
        max_tokens=400,
        system="You are a friendly multimodal assistant. Keep answers concise.",
        messages=messages,
    )
    text = "".join(b.text for b in resp.content if b.type == "text")
    usage = {"input": resp.usage.input_tokens, "output": resp.usage.output_tokens}
    return text, usage


def fake_retrieval(query: str) -> list[str]:
    time.sleep(0.02)
    return [
        f"Memory hit: previously discussed '{query.split()[0]}'",
        "Memory hit: user prefers concise answers",
    ]


def conversation_turn(
    obs: TraceClient,
    llm: Anthropic,
    user_id: str,
    turn: int,
    image_path: Path,
    history: list[dict],
) -> str:
    user_msg = USER_MESSAGES[turn]
    history.append({"role": "user", "text": user_msg})

    with obs.trace(
        name="chat-turn",
        run_type="prod",
        use_case="multimodal-chat",
        metadata={"turn": turn, "session_id": f"sess_{user_id}"},
        tags=["chatbot", "claude"],
    ) as t:
        t.set_user(user_id)
        t.set_metadata(platform="web", model=MODEL)

        try:
            t.attach_image(image_path.read_bytes(), mime="image/png")
        except Exception:
            pass

        with t.step(name="retrieve memory", type="retrieval") as s:
            hits = fake_retrieval(user_msg)
            s.log_message(role="system", text="Searching memory...")
            for h in hits:
                s.log_message(role="system", text=h)
            s.set_metadata(hits=len(hits))

        with t.step(name="claude generate", type="llm", model=MODEL) as s:
            s.log_message(role="user", text=user_msg, images=[image_path.read_bytes()])
            t0 = time.time()
            try:
                reply, usage = call_claude(llm, image_path.read_bytes(), history)
            except Exception as e:
                s.set_status("error")
                s.log_message(role="assistant", text=f"[error] {e!s}")
                raise
            s.log_message(role="assistant", text=reply)
            s.set_token_usage(input=usage["input"], output=usage["output"])
            s.set_metadata(latency_ms=int((time.time() - t0) * 1000))

        if turn == 4:
            t.annotate(label="conversation_complete", score=1.0, user="demo@theta.dev")

    history.append({"role": "assistant", "text": reply})
    return reply


def main() -> None:
    if not IMAGES:
        raise SystemExit(f"No images under {HERE/'images'}/")
    obs = TraceClient(
        api_key=os.environ["THETA_API_KEY"],
        project=os.environ["THETA_PROJECT"],
        base_url=os.environ.get("THETA_BASE_URL", "http://localhost:8080"),
        debug=False,
    )
    llm = Anthropic(api_key=os.environ["ANTHROPIC_API_KEY"])
    user_id = "alice@example.com"
    history: list[dict] = []
    for turn in range(5):
        img = IMAGES[turn % len(IMAGES)]
        print(f"\n-> turn {turn} ({img.name})  user: {USER_MESSAGES[turn]!r}")
        try:
            reply = conversation_turn(obs, llm, user_id, turn, img, history)
            print(f"   claude: {reply.splitlines()[0][:120]}")
        except Exception as e:
            print(f"   ! turn failed: {e!s}")
            break
    obs.flush(timeout=30)
    print("\ndone.")


if __name__ == "__main__":
    random.seed(42)
    main()

File: observability/examples/chatbot-node/chatbot.mjs
/**
 * Multimodal chatbot demo using @theta/observability + real Anthropic Claude.
 *
 * 5 conversation turns where the user uploads an image and asks questions.
 * Each turn = one Trace with 2 Steps (retrieval + llm), image attached.
 *
 * Run:
 *   THETA_API_KEY=tk_... THETA_PROJECT=proj_... THETA_BASE_URL=http://localhost:8080 \
 *   ANTHROPIC_API_KEY=sk-ant-... \
 *   node observability/examples/chatbot-node/chatbot.mjs
 */

import { readFileSync, readdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import Anthropic from "@anthropic-ai/sdk";
import { TraceClient } from "@theta/observability";

const HERE = dirname(fileURLToPath(import.meta.url));
const IMAGES_DIR = join(HERE, "images");
const IMAGES = readdirSync(IMAGES_DIR)
  .filter((f) => f.endsWith(".png"))
  .sort()
  .map((f) => join(IMAGES_DIR, f));

const MODEL = "claude-opus-4-5";
const USER_MESSAGES = [
  "Here's an image — what do you see?",
  "Describe the dominant colors more precisely.",
  "Could this be a sunset, ocean, or forest?",
  "Summarise everything we've discussed so far in two lines.",
  "Write a one-line caption I could use on Instagram.",
];

async function callClaude(llm, imageBytes, history) {
  const imgB64 = Buffer.from(imageBytes).toString("base64");
  const messages = history.map((h) => ({
    role: h.role,
    content: [{ type: "text", text: h.text }],
  }));
  if (messages.length && messages.at(-1).role === "user") {
    messages.at(-1).content.unshift({
      type: "image",
      source: { type: "base64", media_type: "image/png", data: imgB64 },
    });
  }
  const resp = await llm.messages.create({
    model: MODEL,
    max_tokens: 400,
    system: "You are a friendly multimodal assistant. Keep answers concise.",
    messages,
  });
  const text = resp.content
    .filter((b) => b.type === "text")
    .map((b) => b.text)
    .join("");
  return { text, usage: { input: resp.usage.input_tokens, output: resp.usage.output_tokens } };
}

async function fakeRetrieval(query) {
  await new Promise((r) => setTimeout(r, 20));
  return [
    `Memory hit: previously discussed '${query.split(" ")[0]}'`,
    "Memory hit: user prefers concise answers",
  ];
}

async function conversationTurn(obs, llm, userId, turn, imagePath, history) {
  const userMsg = USER_MESSAGES[turn];
  history.push({ role: "user", text: userMsg });
  const imageBytes = readFileSync(imagePath);

  let reply = "";
  await obs.trace(
    {
      name: "chat-turn",
      runType: "prod",
      useCase: "multimodal-chat",
      metadata: { turn, session_id: `sess_${userId}`, model: MODEL, platform: "web" },
      tags: ["chatbot", "claude", "node"],
      userId,
    },
    async (t) => {
      try {
        await t.attachImage(imageBytes, { mime: "image/png" });
      } catch {}

      await t.step({ name: "retrieve memory", type: "retrieval" }, async (s) => {
        const hits = await fakeRetrieval(userMsg);
        s.logMessage({ role: "system", text: "Searching memory..." });
        for (const h of hits) s.logMessage({ role: "system", text: h });
      });

      await t.step({ name: "claude generate", type: "llm", model: MODEL }, async (s) => {
        s.logMessage({ role: "user", text: userMsg, images: [imageBytes] });
        const t0 = Date.now();
        const { text, usage } = await callClaude(llm, imageBytes, history);
        reply = text;
        s.logMessage({ role: "assistant", text });
        s.setTokenUsage({ input: usage.input, output: usage.output });
      });
    }
  );

  history.push({ role: "assistant", text: reply });
  return reply;
}

async function main() {
  if (!IMAGES.length) {
    console.error(`No images under ${IMAGES_DIR}`);
    process.exit(1);
  }
  const obs = new TraceClient({
    apiKey: process.env.THETA_API_KEY,
    project: process.env.THETA_PROJECT,
    baseUrl: process.env.THETA_BASE_URL ?? "http://localhost:8080",
  });
  const llm = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
  const userId = "bob@example.com";
  const history = [];
  for (let turn = 0; turn < 5; turn++) {
    const img = IMAGES[turn % IMAGES.length];
    console.log(`\n-> turn ${turn} (${img.split("/").pop()})  user: ${JSON.stringify(USER_MESSAGES[turn])}`);
    try {
      const reply = await conversationTurn(obs, llm, userId, turn, img, history);
      console.log(`   claude: ${reply.split("\n")[0].slice(0, 120)}`);
    } catch (e) {
      console.error(`   ! turn failed:`, e.message);
      break;
    }
  }
  await obs.flush();
  console.log("\ndone.");
}

main();
